inst/@sym/horzcat.m

```diff
@@ -45,25 +46,8 @@
 
 function h = horzcat(varargin)
 
-  % special case for 0x0 but other empties should be checked for
-  % compatibilty
-  cmd = {
-          '_proc = []'
-          'for i in _ins:'
-          '    if i is None or not i.is_Matrix:'
-          '        _proc.append(sp.Matrix([[i]]))'
-          '    else:'
-          '        if i.shape == (0, 0):'
-          '            pass'
-          '        else:'
-          '            _proc.append(i)'
-          'return sp.MatrixBase.hstack(*_proc),'
-          };
-
-  for i = 1:nargin
-    varargin{i} = sym(varargin{i});
-  end
-  h = pycall_sympy__ (cmd, varargin{:});
+  args = cellfun (@transpose, varargin, 'UniformOutput', false);
+  h = vertcat (args{:}).';
```

Comment:
while I think of it: I approve of this kind of "surface decrease" (having fewer functions that call `pycall_sympy__`) BUT it is important to be aware that there is a downside: namely more "roundtrips" between Python and Octave.  This is most noticeable with `ipc-system` but even with popen2, there are additional copies being made as everything is serialized back-and-forth over the pipe.

Its analogous to HTTP where all objects keep getting converted back-and-forth to JSON.

No action required but its something to keep in mind: perfect software design here might blow-up the latency and IO costs.

Reply:
Interesting thoughts!

Normally when working with a single language, the classic solution is to use an (optimizing) compiler to optimize away all the redundant conversions. Unfortunately, this doesn't work for us because of language barrier: a compiler can't optimize across language boundary :(

---

Some of my thoughts:

First, pythonic may have already solved our problem! While I am not familiar with the internals of pythonic, perhaps using pythonic's `py.`... instead of `pycall_sympy__` can avoid the needless conversions when going through a pipe. Of course, we should benchmark to see if it's indeed the case.

I believe this was the original *future* subproject but we changed it to the current (more urgent) one.

---

Another way is to create a octsmypy python library which contains one python function for each m-file function.

For example, say we want to implement m-file function `@sym/transpose`, `@sym/vertcat` and `@sym/horzcat`. Then our octsmypy python library should contain them as well. More precisely, we can do:

```
from octsympy.sym import transpose, vertcat, horzcat
```

To implement horzcat in terms of vertcat and transpose, we do it inside the octsmypy python library:

```
def horzcat(*XX):
    return transpose(vertcat(*[transpose(X) for X in XX]))
```

without the redundant conversions.
    
Finally, the m-file functions `@sym/transpose`, `@sym/vertcat` and `@sym/horzcat` are now thin wrappers over the corresponding octsympy python library functions, such as:

```
function h = vertcat (varargin)
    args = ...
    h = pycall_sympy__ ('from return octsympy.sym.vertcat(*_ins)', args{:});
end
```

and so on...

Perhaps this can be mentioned in the final report as possible future projects (?)

Reply:
I agree Pythonic offers possibilities.  I think it currently still does conversion: a 'sym' is really just a struct of the 'srepr' (string representation) from SymPy, indep of IPC mechanism.  But your 'xsym' idea would change that.

Agree about mentioing these things ad future projects.

Another thing to mention: CI pythonic runs are twice as slow as Popen2.  I susect there migit be some 'low hanging fruit' in the Pythonic project.

Reply:
And just for the record: another alternative that requires very little effort is (a) being mindful of this issue and aiming for 1 (or a small number) of language crosses per function (b) balancing that against technical debt of code duplication etc.  This is the approach we've been using so far.